src/esthri/src/ops/sync.rs

```diff
@@ -464,12 +488,19 @@ where
                     for entry in entries {
                         let entry = entry.unwrap_object();
                         debug!("key={}", entry.key);
+
+                        let compressed = if transparent_decompress {
+                            head_object_request(s3, bucket, &entry.key).await?.expect("No head info?").is_esthri_compressed()
```

Comment:
Not sure how much overhead this head object request is adding, but I think there is an opportunity for a bit more concurrency here if we need it. Right now we're making one head object request at a time but could do them at the same time, something like

```rust
fn flattened_object_listing<'a, ClientT>(
    s3: &'a ClientT,
    bucket: &'a str,
    key: &'a str,
    directory: &'a Path,
    filters: &'a [GlobFilter],
    transparent_decompress: bool,
) -> impl Stream<Item = Result<(String, Option<ListingMetadata>)>> + 'a
where
    ClientT: S3 + Send + Clone,
{
    async_stream::stream! {
    let mut stream = list_objects_stream(s3, bucket, key);
        loop {
            let entries = match stream.try_next().await {
                Ok(Some(entries)) => entries,
                Ok(None) => break,
                Err(err) => {
                    yield Err(err);
                    return;
                }
            };
            let mut entries = futures::stream::iter(entries)
                .map(|entry| async move {
                    let entry = entry.unwrap_object();
                    debug!("key={}", entry.key);
                    let compressed = if transparent_decompress {
                        head_object_request(s3, bucket, &entry.key)
                            .await?
                            .expect("No head info?")
                            .is_esthri_compressed()
                    } else {
                        false
                    };
                    Ok((entry, compressed))
                })
                .buffered(5);
            loop {
                let (entry, compressed) = match entries.try_next().await {
                    Ok(Some(entry)) => entry,
                    Ok(None) => break,
                    Err(err) => {
                        yield Err(err);
                        return;
                    }
                };
                let path_result = Path::new(&entry.key).strip_prefix(key);
                if let Ok(s3_suffix) = path_result {
                    if process_globs(&s3_suffix, filters).is_some() {
                        let local_path: String = directory.join(&s3_suffix).to_string_lossy().into();
                        let s3_suffix = s3_suffix.to_string_lossy().into();
                        yield Ok((
                            local_path,
                            ListingMetadata::some(s3_suffix, entry.e_tag, compressed),
                        ));
                    }
                } else {
                    yield Err(path_result.err().unwrap().into());
                    return;
                }
            }
        }
    }
}
```
Not 100% sure that will work but the main point is to process all the head requests for each batch of entries at the same time

Reply:
Aha, this is great, much nicer! Will definitely speed up syncs with lots of files.